### DISCUSSION TITLE
How to access (and update) values in component *after* using FieldArray helpers

### DISCUSSION THREAD

**Discussion by genovese** [jaredpalmer/formik#3116] [Q&A] [answered]
After removing an element of a field array with the remove helper, I want to adjust some metadata in the other array elements, but there appears to be no way to access the updated values from within the handler.  (Any `values` obtained from `useFormikContext`, for instance, refers to the wrong object after using `helpers.remove`.)

So to clarify, here's a schematic of the problem

```
const Component = ({index, helpers}) => {
    const { values, setValues } = useFormikContext();
    ...
    deleteEntryHandler = () => {
        helpers.remove(index);
        setValues(updateStuffInRemainingEntries(???));  // values doesn't work here; it no longer points to the right state
        ...
    }
    ...
    return  <div> ...    <YadaYada onClick={deleteEntryHandler} /> ... </div>;
}
```
Here, `updateStuffInRemainingEntries` might say, use immer and the current `values` to return an updated version that is passed to `setValues`.

An obvious approach would be to just handle the remove directly instead of using `helpers.remove`, and I can do that. But it seems a shame to lose the benefits of the pre-canned helpers (which handle touched and error nicely, for instance) and have to reimplement that.

Is there a way to access an _updated_ state (i.e., `values`) inside a component/handler after using the FieldArray helpers?

Thanks

**Comment by johnrom** [comment, marked as answer]
In v2 you need to use an effect for this because it relies solely on hooks. In v3 we will most likely provide an escape hatch in the form of `getState()` to get the latest state in "the future".

```js
    const { values, setValues } = useFormikContext();
    // deleteEntryHandler etc
    // ...
    useEffect(() => {
      setValues(updateStuffInRemainingEntries(values.yourArrayFieldName));
    }, [values.yourArrayFieldName]);

    return  <div><YadaYada onClick={deleteEntryHandler} /></div>;
```

**Reply by johnrom** [reply]
setValues now accepts a callback, so you can do something like: 

```js
setValues(values => getRemainingEntriesAfterUpdate(values));
```

**Reply by genovese** [reply]
Thanks, that's very helpful and makes sense.  This does separate the two changes (deletion and subsequent update of the internal data), so things are out of sync for part of a cycle. But that seems ok....  getState would be helpful. Thanks again.
